File: F/3043-find-the-length-of-the-longest-common-prefix/main.go
package main

type Pre struct {
	Child []*Pre
}

func longestCommonPrefix(arr1 []int, arr2 []int) int {
	var build func(d []int, idx int, pre *Pre)
	build = func(d []int, idx int, pre *Pre) {
		if idx == -1 {
			return
		}

		if pre.Child[d[idx]] == nil {
			pre.Child[d[idx]] = &Pre{Child: make([]*Pre, 10)}
		}
		build(d, idx-1, pre.Child[d[idx]])
	}

	var search func(d []int, idx int, step int, pre *Pre, ans *int)
	search = func(d []int, idx int, step int, pre *Pre, ans *int) {
		if idx == -1 || pre.Child[d[idx]] == nil {
			if step > *ans {
				*ans = step
			}
		} else {
			search(d, idx-1, step+1, pre.Child[d[idx]], ans)
		}
	}

	root := &Pre{Child: make([]*Pre, 10)}
	for _, v := range arr1 {
		d := []int{}
		vv := v
		for vv > 0 {
			d = append(d, vv%10)
			vv /= 10
		}

		build(d, len(d)-1, root)
	}

	ans := 0
	for _, v := range arr2 {
		d := []int{}
		vv := v
		for vv > 0 {
			d = append(d, vv%10)
			vv /= 10
		}

		search(d, len(d)-1, 0, root, &ans)
	}
	return ans

}

File: F/3043-find-the-length-of-the-longest-common-prefix/main.rs
use std::collections::HashMap;

impl Solution {
  pub fn longest_common_prefix(arr1: Vec<i32>, arr2: Vec<i32>) -> i32 {
    let mut dm: HashMap<i32, i32> = HashMap::new();
    arr1.iter().for_each(|&v| {
      let mut vv = v;
      let mut l = v.checked_ilog10().unwrap_or(0) as i32 + 1;
      while vv > 0 {
        dm.insert(vv, l);
        vv /= 10;
        l -= 1;
      }
    });

    let mut ans: i32 = 0;
    arr2.iter().for_each(|&v| {
      let mut vv = v;
      let mut l = v.checked_ilog10().unwrap_or(0) as i32 + 1;
      while vv > 0 {
        if dm.contains_key(&vv) {
          ans = (*dm.get(&vv).unwrap()).max(ans);
          return;
        }
        vv /= 10;
      }
    });
    ans
  }
}
